### PR TITLE
Fix implicit casts for dimension and various other related things.

### PR DESCRIPTION
**Pull request by tex3d**
- Fix improper bool early promotion to int in CheckBinOpForHLSL
- Modify VerifierTest.cpp to re-enable parsing by VerifierHelper.py
- Update/fix VerifierHelper.py
- Update/fix various tests, including adding expected fxc comments
- Fix source locations for ValidateCast uses
- ValidateCast will use expression location if no loc provided
- TODO: Fix error on explicit truncation cast that should be ok, as long
  as it's not an l-value "cannot truncate lvalue vector/matrix"
  See binop-dims.hlsl
- TODO: Fix L-Value casts in the future

### CODE REVIEW

**Comment by tex3d** [comment]
FYI @antiagainst, @ehsannas: This test appears to break CodeGenSPIRV test, but I think test may need to be updated.  When I look at the AST diff for the test, there are two kinds of changes:
1. Binary operator preserves typedef type in result if it came from one of the sides, for example preventing it from always dropping friendly 'float4' in favor of 'vector<float, 4>', which can be better for diagnostics.  The log on AppVeyor appears to indicate a case with this kind of change isn't matching.
2. Given: `float f; float1 f1;`  ...  Now: `f1 = f1 * f;` results in cast of f to float1, and bin op '*' type is float1, rather than cast of f1 to scalar float, then bin op '*' in float, then cast back to float1 for assignment.  I think the new sequence is more consistent with general vector * scalar behavior, rather than appearing to be a special case for when the vector is size 1.  I'm not sure if this is causing a failure, since the log only appears to capture the first failure.

**Comment by ehsannas** [comment]
Thanks for the heads up @tex3d . I'll look into the SPIR-V side and push a fix into your pull-request.

**Comment by ehsannas** [comment]
@antiagainst PTAL

**Comment by ehsannas** [comment]
A couple of HLSL tests failed on Appveyor, but didn't fail locally for me. I have pushed a merge commit to bring this branch to TOT (the branch was 6 days old). Hopefully this should fix the tests failures.

**Comment by tex3d** [comment]
Note: the following AppVeyor failure is a known issue that is failing sometimes, that I had been looking into the cause of.  So for now, we should ignore this particular failure pattern.
CompilerTest::QuickTest
c:\projects\DirectXShaderCompiler\tools\clang\test\HLSL\..\CodeGenHLSL\quick-test\d3dreflect\lib_hs_export1.hlsl

File: tools/clang/lib/Sema/SemaHLSL.cpp [hunk 1]
```diff
@@ -7490,183 +7484,22 @@ bool HandleVoidConversion(QualType source, QualType target, bool explicitConvers
...
+  // 3. A vector may be truncated to a smaller vector.
+  // 4. A matrix may be truncated to a smaller matrix.
+  // 5. The result of a vector and a matrix is:
   //    a. If the matrix has one row it's a vector-sized
   //       piece of the row.
   //    b. If the matrix has one column it's a vector-sized
   //       piece of the column.
   //    c. Otherwise the number of elements in the vector
   //       and matrix must match and the result is the vector.
   // 5. The result of a matrix and a vector is similar to #4.
```
[the remark this answers is not in the document]
Comment (tex3d) [reply tools/clang/lib/Sema/SemaHLSL.cpp]: Pushed commit that addresses these comment issues.

**Pull request merged**